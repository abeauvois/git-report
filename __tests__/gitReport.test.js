const { from, of, zip, pipe, combineLatest } = require("rxjs");
const {
  toArray,
  skip,
  map,
  take,
  mergeMap,
  tap,
  groupBy,
  merge,
  mergeAll,
  concatMap,
  expand,
  filter,
  flatMap,
} = require("rxjs/operators");
const { fromCSVFile } = require("../src/fromCSVFile");
const {
  toDateTimeMapper,
  groupByYear,
  groupByMonth,
  groupByDay,
  groupByAmPm,
  mergeByGroup,
  // groupMessagesByAmPm,
  groupAndMergeByMonth,
  groupAndMergeByDay,
  groupMessagesByDay,
  buildReport,
  sendGitReport,
} = require("../src/report");

describe("gitReport", () => {
  const filename = "./__tests__/test_file.csv";

  test("R1: fromCSVFile", (done) => {
    fromCSVFile(filename)
      .pipe(toArray())
      .subscribe(
        (received) => {
          expect(received).toMatchSnapshot();
          done();
        },
        (e) => console.log("Errroor", e)
      );
  });

  test("R2: toDateTimeMapper", (done) => {
    fromCSVFile(filename)
      .pipe(skip(1), take(1), map(toDateTimeMapper))
      .subscribe(
        (received) => {
          expect(received).toMatchSnapshot();
          done();
        },
        (e) => console.log("Errroor", e)
      );
  });

  test("R3.a: groupByYear & mergeByGroup", (done) => {
    fromCSVFile(filename)
      .pipe(skip(1), map(toDateTimeMapper), groupByYear, mergeByGroup)
      .subscribe(
        (received) => {
          expect(received).toMatchSnapshot();
          done();
        },
        (e) => console.log("Errroor", e)
      );
  });

  test("R3.b: groupByMonth & mergeByGroup", (done) => {
    fromCSVFile(filename)
      .pipe(skip(1), map(toDateTimeMapper), groupByMonth, mergeByGroup)
      .subscribe(
        (received) => {
          expect(received).toMatchSnapshot();
          done();
        },
        (e) => console.log("Errroor", e)
      );
  });

  test("R4: groupAndMergeByMonth", (done) => {
    fromCSVFile(filename)
      .pipe(skip(1), map(toDateTimeMapper), groupAndMergeByMonth)
      .subscribe(
        (received) => {
          expect(received).toMatchSnapshot();
          done();
        },
        (e) => console.log("Errroor", e)
      );
  });

  test("R5: groupMessagesByDay", (done) => {
    fromCSVFile(filename)
      .pipe(skip(1), map(toDateTimeMapper), groupAndMergeByDay, mergeMap(groupMessagesByDay))
      .subscribe(
        (received) => {
          expect(received).toMatchSnapshot();
          done();
        },
        (e) => console.log("Errroor", e)
      );
  });

  test("R6: groupByMonthByDay", (done) => {
    const asArray = (group) => group.pipe(toArray());
    const groupMapper = (group) => zip(of(group.key), asArray(group));
    const mergeGroup = mergeMap(groupMapper);

    const source$ = fromCSVFile(filename).pipe(skip(1), map(toDateTimeMapper));

    const sourceByYear$ = source$.pipe(groupByYear, mergeGroup);
    const sourceByMonth$ = (list) => from(list).pipe(groupByMonth, mergeGroup);
    const sourceByDay$ = (list) => from(list).pipe(groupByDay, mergeGroup);

    sourceByYear$
      .pipe(
        filter(([_, list]) => Boolean(list)), // remove undefined list
        flatMap(([year, list]) => combineLatest([of(year), sourceByMonth$(list)])),
        flatMap(([year, [month, list]]) => combineLatest([of(year), combineLatest([of(month), sourceByDay$(list)])])),
        flatMap(([year, [month, [day, list]]]) =>
          combineLatest([of(year), combineLatest([of(month), groupMessagesByDay([day, list])])])
        )
        // tap((t) => console.log("t:", t))
      )
      .subscribe(
        (received) => {
          expect(received).toMatchSnapshot();
          done();
        },
        (e) => console.log("Errroor", e)
      );
  });

  test("R7: buildReport", (done) => {
    buildReport({ filename }).then((received) => {
      expect(received).toMatchSnapshot();
      done();
    });
  });

  test("R8: sendGitReport to console", (done) => {
    sendGitReport({ filename: filename, channel: "console" }).then((received) => {
      expect(received).toMatchSnapshot();
      done();
    });
  });
});
