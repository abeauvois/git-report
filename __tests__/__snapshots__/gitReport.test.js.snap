// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gitReport R1: fromCSVFile 1`] = `
Array [
  Object {
    "contributor": " contributor",
    "date": " date",
    "message": " message",
    "sha": "sha",
  },
  Object {
    "contributor": " abeauvois",
    "date": " Fri Nov 13 10:40:39 2020",
    "message": " \\"m1 13 nov\\"",
    "sha": "2",
  },
  Object {
    "contributor": " abeauvois",
    "date": " Thu Nov 12 16:06:45 2020",
    "message": " \\"m2 12 nov PM\\"",
    "sha": "3",
  },
  Object {
    "contributor": " abeauvois",
    "date": " Thu Nov 12 10:03:21 2020",
    "message": " \\"m3 12 nov AM\\"",
    "sha": "4",
  },
  Object {
    "contributor": " abeauvois",
    "date": " Thu Nov 12 11:03:21 2020",
    "message": " \\"m3 12 nov AM\\"",
    "sha": "4",
  },
  Object {
    "contributor": " abeauvois",
    "date": " Thu Oct 12 16:03:21 2020",
    "message": " \\"m2 oct\\"",
    "sha": "5",
  },
  Object {
    "contributor": " abeauvois",
    "date": " Thu Sep 12 16:03:21 2020",
    "message": " \\"m1 sep\\"",
    "sha": "6",
  },
  Object {
    "contributor": " abeauvois",
    "date": " Fri Jan 01 10:40:39 2021",
    "message": " \\"m1 01 janv 21\\"",
    "sha": "1",
  },
  Object {
    "contributor": " abeauvois",
    "date": " Mon Feb 01 13:40:39 2021",
    "message": " \\"m1 01 fev 21\\"",
    "sha": "0",
  },
  Object {
    "contributor": " abeauvois",
    "date": " Wed Nov 03 13:40:39 2021",
    "message": " \\"m1 03 nov 21\\"",
    "sha": "0",
  },
]
`;

exports[`gitReport R2: toDateTimeMapper 1`] = `
Object {
  "ampm": "am",
  "day": 13,
  "dayWeek": 5,
  "hour": 10,
  "message": "m1 13 nov",
  "month": 10,
  "year": 2020,
}
`;

exports[`gitReport R3.a: groupByYear & mergeByGroup 1`] = `
Array [
  2020,
  Array [
    Object {
      "ampm": "am",
      "day": 13,
      "dayWeek": 5,
      "hour": 10,
      "message": "m1 13 nov",
      "month": 10,
      "year": 2020,
    },
    Object {
      "ampm": "pm",
      "day": 12,
      "dayWeek": 4,
      "hour": 16,
      "message": "m2 12 nov PM",
      "month": 10,
      "year": 2020,
    },
    Object {
      "ampm": "am",
      "day": 12,
      "dayWeek": 4,
      "hour": 10,
      "message": "m3 12 nov AM",
      "month": 10,
      "year": 2020,
    },
    Object {
      "ampm": "am",
      "day": 12,
      "dayWeek": 4,
      "hour": 11,
      "message": "m3 12 nov AM",
      "month": 10,
      "year": 2020,
    },
    Object {
      "ampm": "pm",
      "day": 12,
      "dayWeek": 1,
      "hour": 16,
      "message": "m2 oct",
      "month": 9,
      "year": 2020,
    },
    Object {
      "ampm": "pm",
      "day": 12,
      "dayWeek": 6,
      "hour": 16,
      "message": "m1 sep",
      "month": 8,
      "year": 2020,
    },
  ],
]
`;

exports[`gitReport R3.a: groupByYear & mergeByGroup 2`] = `
Array [
  2021,
  Array [
    Object {
      "ampm": "am",
      "day": 1,
      "dayWeek": 5,
      "hour": 10,
      "message": "m1 01 janv 21",
      "month": 0,
      "year": 2021,
    },
    Object {
      "ampm": "pm",
      "day": 1,
      "dayWeek": 1,
      "hour": 13,
      "message": "m1 01 fev 21",
      "month": 1,
      "year": 2021,
    },
    Object {
      "ampm": "pm",
      "day": 3,
      "dayWeek": 3,
      "hour": 13,
      "message": "m1 03 nov 21",
      "month": 10,
      "year": 2021,
    },
  ],
]
`;

exports[`gitReport R3.b: groupByMonth & mergeByGroup 1`] = `
Array [
  10,
  Array [
    Object {
      "ampm": "am",
      "day": 13,
      "dayWeek": 5,
      "hour": 10,
      "message": "m1 13 nov",
      "month": 10,
      "year": 2020,
    },
    Object {
      "ampm": "pm",
      "day": 12,
      "dayWeek": 4,
      "hour": 16,
      "message": "m2 12 nov PM",
      "month": 10,
      "year": 2020,
    },
    Object {
      "ampm": "am",
      "day": 12,
      "dayWeek": 4,
      "hour": 10,
      "message": "m3 12 nov AM",
      "month": 10,
      "year": 2020,
    },
    Object {
      "ampm": "am",
      "day": 12,
      "dayWeek": 4,
      "hour": 11,
      "message": "m3 12 nov AM",
      "month": 10,
      "year": 2020,
    },
    Object {
      "ampm": "pm",
      "day": 3,
      "dayWeek": 3,
      "hour": 13,
      "message": "m1 03 nov 21",
      "month": 10,
      "year": 2021,
    },
  ],
]
`;

exports[`gitReport R3.b: groupByMonth & mergeByGroup 2`] = `
Array [
  9,
  Array [
    Object {
      "ampm": "pm",
      "day": 12,
      "dayWeek": 1,
      "hour": 16,
      "message": "m2 oct",
      "month": 9,
      "year": 2020,
    },
  ],
]
`;

exports[`gitReport R3.b: groupByMonth & mergeByGroup 3`] = `
Array [
  8,
  Array [
    Object {
      "ampm": "pm",
      "day": 12,
      "dayWeek": 6,
      "hour": 16,
      "message": "m1 sep",
      "month": 8,
      "year": 2020,
    },
  ],
]
`;

exports[`gitReport R3.b: groupByMonth & mergeByGroup 4`] = `
Array [
  0,
  Array [
    Object {
      "ampm": "am",
      "day": 1,
      "dayWeek": 5,
      "hour": 10,
      "message": "m1 01 janv 21",
      "month": 0,
      "year": 2021,
    },
  ],
]
`;

exports[`gitReport R3.b: groupByMonth & mergeByGroup 5`] = `
Array [
  1,
  Array [
    Object {
      "ampm": "pm",
      "day": 1,
      "dayWeek": 1,
      "hour": 13,
      "message": "m1 01 fev 21",
      "month": 1,
      "year": 2021,
    },
  ],
]
`;

exports[`gitReport R4: groupAndMergeByMonth 1`] = `
Array [
  10,
  Array [
    Object {
      "ampm": "am",
      "day": 13,
      "dayWeek": 5,
      "hour": 10,
      "message": "m1 13 nov",
      "month": 10,
      "year": 2020,
    },
    Object {
      "ampm": "pm",
      "day": 12,
      "dayWeek": 4,
      "hour": 16,
      "message": "m2 12 nov PM",
      "month": 10,
      "year": 2020,
    },
    Object {
      "ampm": "am",
      "day": 12,
      "dayWeek": 4,
      "hour": 10,
      "message": "m3 12 nov AM",
      "month": 10,
      "year": 2020,
    },
    Object {
      "ampm": "am",
      "day": 12,
      "dayWeek": 4,
      "hour": 11,
      "message": "m3 12 nov AM",
      "month": 10,
      "year": 2020,
    },
    Object {
      "ampm": "pm",
      "day": 3,
      "dayWeek": 3,
      "hour": 13,
      "message": "m1 03 nov 21",
      "month": 10,
      "year": 2021,
    },
  ],
]
`;

exports[`gitReport R4: groupAndMergeByMonth 2`] = `
Array [
  9,
  Array [
    Object {
      "ampm": "pm",
      "day": 12,
      "dayWeek": 1,
      "hour": 16,
      "message": "m2 oct",
      "month": 9,
      "year": 2020,
    },
  ],
]
`;

exports[`gitReport R4: groupAndMergeByMonth 3`] = `
Array [
  8,
  Array [
    Object {
      "ampm": "pm",
      "day": 12,
      "dayWeek": 6,
      "hour": 16,
      "message": "m1 sep",
      "month": 8,
      "year": 2020,
    },
  ],
]
`;

exports[`gitReport R4: groupAndMergeByMonth 4`] = `
Array [
  0,
  Array [
    Object {
      "ampm": "am",
      "day": 1,
      "dayWeek": 5,
      "hour": 10,
      "message": "m1 01 janv 21",
      "month": 0,
      "year": 2021,
    },
  ],
]
`;

exports[`gitReport R4: groupAndMergeByMonth 5`] = `
Array [
  1,
  Array [
    Object {
      "ampm": "pm",
      "day": 1,
      "dayWeek": 1,
      "hour": 13,
      "message": "m1 01 fev 21",
      "month": 1,
      "year": 2021,
    },
  ],
]
`;

exports[`gitReport R5: groupMessagesByDay 1`] = `
Object {
  "day": 13,
  "dayWeek": 5,
  "messagesAm": " m1 13 nov",
  "messagesPm": "",
}
`;

exports[`gitReport R5: groupMessagesByDay 2`] = `
Object {
  "day": 12,
  "dayWeek": 4,
  "messagesAm": " m3 12 nov AM, m3 12 nov AM",
  "messagesPm": " m2 12 nov PM, m2 oct, m1 sep",
}
`;

exports[`gitReport R5: groupMessagesByDay 3`] = `
Object {
  "day": 1,
  "dayWeek": 5,
  "messagesAm": " m1 01 janv 21",
  "messagesPm": " m1 01 fev 21",
}
`;

exports[`gitReport R5: groupMessagesByDay 4`] = `
Object {
  "day": 3,
  "dayWeek": 3,
  "messagesAm": "",
  "messagesPm": " m1 03 nov 21",
}
`;

exports[`gitReport R6: groupByMonthByDay 1`] = `
Array [
  2020,
  Array [
    10,
    Object {
      "day": 13,
      "dayWeek": 5,
      "messagesAm": " m1 13 nov",
      "messagesPm": "",
    },
  ],
]
`;

exports[`gitReport R6: groupByMonthByDay 2`] = `
Array [
  2020,
  Array [
    10,
    Object {
      "day": 12,
      "dayWeek": 4,
      "messagesAm": " m3 12 nov AM, m3 12 nov AM",
      "messagesPm": " m2 12 nov PM",
    },
  ],
]
`;

exports[`gitReport R6: groupByMonthByDay 3`] = `
Array [
  2020,
  Array [
    9,
    Object {
      "day": 12,
      "dayWeek": 1,
      "messagesAm": "",
      "messagesPm": " m2 oct",
    },
  ],
]
`;

exports[`gitReport R6: groupByMonthByDay 4`] = `
Array [
  2020,
  Array [
    8,
    Object {
      "day": 12,
      "dayWeek": 6,
      "messagesAm": "",
      "messagesPm": " m1 sep",
    },
  ],
]
`;

exports[`gitReport R6: groupByMonthByDay 5`] = `
Array [
  2021,
  Array [
    0,
    Object {
      "day": 1,
      "dayWeek": 5,
      "messagesAm": " m1 01 janv 21",
      "messagesPm": "",
    },
  ],
]
`;

exports[`gitReport R6: groupByMonthByDay 6`] = `
Array [
  2021,
  Array [
    1,
    Object {
      "day": 1,
      "dayWeek": 1,
      "messagesAm": "",
      "messagesPm": " m1 01 fev 21",
    },
  ],
]
`;

exports[`gitReport R6: groupByMonthByDay 7`] = `
Array [
  2021,
  Array [
    10,
    Object {
      "day": 3,
      "dayWeek": 3,
      "messagesAm": "",
      "messagesPm": " m1 03 nov 21",
    },
  ],
]
`;

exports[`gitReport R7: buildReport 1`] = `
"[34mDec[39m 2020
[33mFri[39m, [33m13[39m
AM: [33m m1 13 nov[39m
PM: 

[34mDec[39m 2020
[33mThu[39m, [33m12[39m
AM: [33m m3 12 nov AM, m3 12 nov AM[39m
PM: [33m m2 12 nov PM[39m

[34mNov[39m 2020
[33mMon[39m, [33m12[39m
AM: 
PM: [33m m2 oct[39m

[34mOct[39m 2020
[33mSat[39m, [33m12[39m
AM: 
PM: [33m m1 sep[39m

[34mJan[39m 2021
[33mFri[39m, [33m1[39m
AM: [33m m1 01 janv 21[39m
PM: 

[34mFeb[39m 2021
[33mMon[39m, [33m1[39m
AM: 
PM: [33m m1 01 fev 21[39m

[34mDec[39m 2021
[33mWed[39m, [33m3[39m
AM: 
PM: [33m m1 03 nov 21[39m

"
`;

exports[`gitReport R8: sendGitReport to console 1`] = `
"[34mDec[39m 2020
[33mFri[39m, [33m13[39m
AM: [33m m1 13 nov[39m
PM: 

[34mDec[39m 2020
[33mThu[39m, [33m12[39m
AM: [33m m3 12 nov AM, m3 12 nov AM[39m
PM: [33m m2 12 nov PM[39m

[34mNov[39m 2020
[33mMon[39m, [33m12[39m
AM: 
PM: [33m m2 oct[39m

[34mOct[39m 2020
[33mSat[39m, [33m12[39m
AM: 
PM: [33m m1 sep[39m

[34mJan[39m 2021
[33mFri[39m, [33m1[39m
AM: [33m m1 01 janv 21[39m
PM: 

[34mFeb[39m 2021
[33mMon[39m, [33m1[39m
AM: 
PM: [33m m1 01 fev 21[39m

[34mDec[39m 2021
[33mWed[39m, [33m3[39m
AM: 
PM: [33m m1 03 nov 21[39m

"
`;
